File: curso-noob-poo-typescript/animal.ts
export default class Animal {
  private tipo: string;
  private nome: string;
  private idade: number;

  constructor(t: string, n: string, i: number){
    this.tipo = t;
    this.nome = n;
    this.idade = i;
  }

  protected comunicar(): void{
    console.log("Olá, sou método da classe mãe")
  }
}

class Cachorro extends Animal {
  raça: string

  constructor(t: string, n: string, i: number, r: string){
    super(t, n, i)
    this.raça = r;
  }

  public latir(){
    super.comunicar()
  }
}

class Gente extends Animal {
  cor: string;

  constructor(t: string, n: string, i: number, c: string){
    super(t, n, i)
    this.cor = c;
  }  
}

let c1 = new Cachorro("cachorro", "bob", 2, 'poodle')
c1.latir()

let g1 = new Gente("Humano", "Marley", 25, 'Negro')
// g1.comunicar()

File: curso-noob-poo-ruby/classe/animal.rb
require_relative "modulo"

class Animal
  include Comunica
  attr_reader :nome

  def initialize(nome)
    @nome = nome
  end
end

class Cachorro < Animal
  attr_reader :raça

  def initialize(nome, raça)
    super(nome)
    @raça = raça
  end

  def comunicar
    puts "Au, au, au! = Eu sou a classe filha Cachorro"
  end
end

class Gato < Animal
  attr_reader :raça

  def initialize(nome, raça)
    super(nome)
    @raça = raça
  end

  def comunicar
    puts "Miau, miau! = Eu sou a classe filha Gato"
  end
end

bob = Cachorro.new("Bob", "Poodle")
puts bob.nome
puts bob.raça
bob.comunicar
puts "-----"
xaninho = Gato.new("Xaninho", "Siamês")
puts xaninho.nome
puts xaninho.raça
xaninho.comunicar